### PR TITLE
kubeadm: simplify minimum Kubernetes version bumps

### PR DESCRIPTION
**Pull request by bart0sh**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Replaced hardcoded "v0.12.0" strings with new
MinimalKubernetesVersion constant.

This should help with a regular release version bumps.

**Special notes for your reviewer**:

This is a first step towards fixing kubernetes/kubeadm#1260

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion_test.go
Comment (rosti) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion_test.go]: We don't need this one. The tested function does not depend on the current Kubernetes version.

File: cmd/kubeadm/app/cmd/phases/util_test.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/phases/util_test.go]: Again, we don't need this to test `SetKubernetesVersion`. It can stay forever on v1.12.0.

File: cmd/kubeadm/app/componentconfigs/config_test.go
Comment (rosti) [inline cmd/kubeadm/app/componentconfigs/config_test.go]: I don't think, that this is necessary either.

File: cmd/kubeadm/app/util/version_test.go
Comment (rosti) [inline cmd/kubeadm/app/util/version_test.go]: `MinimumKubernetesVersion` is not needed to test the logic in `validateStableVersion`

File: cmd/kubeadm/app/constants/constants.go
Comment (rosti) [inline cmd/kubeadm/app/constants/constants.go]: We can probably trim this to `v1.12`, so we can cover more cases with this PR (like `v1.12.5`).

**Review by rosti — COMMENTED** [review]
Thanks for taking this up @bart0sh !
We can cover a few more cases with this PR (all `v1.12.x` cases).
A few of the spots don't depend on the current version. It's OK to keep them wired to a string as that may make code more readable.

File: cmd/kubeadm/app/constants/constants.go
Comment (bart0sh) [reply cmd/kubeadm/app/constants/constants.go]: I'd prefer to do it in a follow-up PR as this one is already big and could cause rebase conflicts if not merged soon. wdyh?

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion_test.go
Comment (bart0sh) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion_test.go]: Thank you for the review. will revert this and all the following cases you've pointed out.

File: cmd/kubeadm/app/constants/constants.go
Comment (neolit123) [reply cmd/kubeadm/app/constants/constants.go]: better to have it in the same PR, IMO.

Comment (bart0sh) [reply cmd/kubeadm/app/constants/constants.go]: ok, will do

Comment (neolit123) [reply cmd/kubeadm/app/constants/constants.go]: but i wonder what other cases we need to do?

**Review by neolit123 — APPROVED** [review]
i think this is good.
thanks @bart0sh

File: cmd/kubeadm/app/constants/constants.go
Comment (bart0sh) [reply cmd/kubeadm/app/constants/constants.go]: @rosti, @neolit123 Would something like this be acceptable:
```
// MinK8sVerMajorMinor specifies <major>.<minor> part of the minimum Kubernetes version supported by Kubeadm

MinK8sMajorMinor = "v1.12"

// MinimumKubernetesVersion specifies minimum Kubernetes version supported by kubeadm
MinimumKubernetesVersion = MinK8sVerMajorMinor + ".0"
```

Then we can use MinK8sVerMajorMinor constant in the rest of v1.12* strings.

Comment (neolit123) [reply cmd/kubeadm/app/constants/constants.go]: on each version bump we need to send a PR that does something like this:
https://github.com/kubernetes/kubernetes/pull/68649/files

the idea is to reduce this to a change in a single file - constants.go.

Comment (bart0sh) [reply cmd/kubeadm/app/constants/constants.go]: I think I understood the idea. I don't understand why it has to be implemented as one PR.

Comment (neolit123) [reply cmd/kubeadm/app/constants/constants.go]: my vote would be for:
```
MinimumKubernetesVersionWithoutPatch = ...
MinimumKubernetesVersion = ...
```

in terms of the example change in:
https://github.com/kubernetes/kubernetes/pull/68649/files

we might as well add more constants for the sake of making it more easy.
e.g. Minor + 1

that's my 2c.

**Comment by neolit123** [comment]
/approve
/hold
/assign @fabriziopandini 
for another set of eyes.

File: cmd/kubeadm/app/constants/constants.go
Comment (rosti) [reply cmd/kubeadm/app/constants/constants.go]: I was thinking about having all changes, that involve `1.12.x` in a single PR (currently we replaced only `1.12.0` and left some `1.12.x` stuff, where x != 0, for another PR).

I am OK with moving `1.13` and `1.14` stuff in other PR(s). That way it will be more reviewable.

Comment (bart0sh) [reply cmd/kubeadm/app/constants/constants.go]: @rosti 

> currently we replaced only 1.12.0 and left some 1.12.x stuff

Nope, almost everything is covered with MinimumKubernetesVersion and MinimumKubernetesVersionWithoutPatch. You probably missed my updates.

Below is what's left currently. Can you point me out what of this should be included into this PR?
```
$ grep -r v1.12 . | grep -v ':\s*\/\/'
Binary file ./kubeadm matches
./app/constants/constants.go:	MinimumKubernetesVersionWithoutPatch = "v1.12"
Binary file ./app/cmd/phases/phases.test matches
./app/cmd/phases/util_test.go:			input:  "v1.12.0",
./app/cmd/phases/util_test.go:			output: "v1.12.0",
./app/cmd/config_test.go:				":v1.12.1",
./app/cmd/config_test.go:				kubernetesVersion: v1.12.1
./app/componentconfigs/config_test.go:	k8sVersion := version.MustParseGeneric("v1.12.0")
./app/phases/uploadconfig/uploadconfig_test.go:					KubernetesVersion: "v1.12.10",
Binary file ./app/util/upgrade/upgrade.test matches
./app/util/config/testdata/conversion/master/v1alpha3.yaml:kubernetesVersion: v1.12.2
./app/util/config/testdata/conversion/master/v1alpha3.yaml:unifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2"
./app/util/config/testdata/conversion/master/v1beta1.yaml:kubernetesVersion: v1.12.2
./app/util/config/testdata/conversion/master/internal.yaml:KubernetesVersion: v1.12.2
./app/util/config/common.go:		"kubeadm.k8s.io/v1alpha2": "v1.12",
./app/util/version_test.go:			remoteVersion: "v1.12.0",
./app/util/version_test.go:			remoteVersion: "v1.12.1",
./app/util/version_test.go:			output:        "v1.12.1",
./app/util/version_test.go:			clientVersion: "v1.12.0",
./app/util/version_test.go:			remoteVersion: "v1.12.0",
./app/util/etcd/etcd.go:				klog.V(1).Infoln("etcd manifest created by kubeadm v1.12")
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.0",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/hyperkube:v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				KubernetesVersion:        "v1.12.2",
./app/apis/kubeadm/v1alpha3/conversion_test.go:				UnifiedControlPlaneImage: "k8s.gcr.io/otherimage:v1.12.2",
```

Comment (rosti) [reply cmd/kubeadm/app/constants/constants.go]: Sorry @bart0sh , I missed the updates. Will take a look.

**Review by rosti — APPROVED** [review]
Thanks @bart0sh ! Just a single one missing:
https://github.com/kubernetes/kubernetes/blob/1c04caa04325e1f64d9a15714ad61acdd2a81013/cmd/kubeadm/test/cmd/init_test.go#L105

Then we are ready to LGTM the 1.12.x portion.

**Comment by bart0sh** [comment]
@rosti Thank you for the review. Updated. lgtm, please as I'm concerned about possible conflicts.

**Comment by rosti** [comment]
Thanks again @bart0sh !
/lgtm
/test pull-kubernetes-e2e-kops-aws

**Review by fabriziopandini — COMMENTED** [review]
@bart0sh 
Thanks for this PR! simplify Kubernetes version bumps will be of great help!

One background question for this PR. Have you considered to implement a more declarative approach for replacing all the "version string concatenation magic" currently existing in kubeadm tests? e.g.
 
```
func (v Version) WithMinor(patch uint) *Version {
}

func (v Version) WithPatch(patch uint) *Version {
}

func (v Version) WithPreRelease(prerelease string) *Version {
}
```

Or something similar for getting variations of the `MinimumKubernetesVersion `?

**Comment by bart0sh** [comment]
@fabriziopandini Thank you for the idea. So far I don't see that ```constants.MinimumKubernetesVersion + "-alpha.1``` is much worse than  ```constants.MinimumKubernetesVersion.WithPatch(0-alpha.1)```, but can play with it to see if it makes sense. 

BTW, does anybody know about existing Kubernetes APIs in this area?

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/alpha/kubelet.go]: No need for the `.String()` portion here, `Sprintf` will do it for you.

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/alpha/kubelet.go]: And here.

File: cmd/kubeadm/app/cmd/config_test.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/config_test.go]: And here.

File: cmd/kubeadm/test/util.go
Comment (rosti) [inline cmd/kubeadm/test/util.go]: And here.

**Comment by bart0sh** [comment]
@neolit123, @fabriziopandini, @rosti I've replaced string concatenations with using With* methods as @fabriziopandini suggested. Please, review again. Thanks.

**Comment by neolit123** [comment]
@bart0sh 
the `staging/src/k8s.io/apimachinery/pkg/util/version/version.go` helper changes LGTM, but FYI this can drag this PR a lot in time as it enters the scope of another SIG.

**Review by rosti — APPROVED** [review]
Thanks for the update @bart0sh !
Actually I like this approach better, but as @neolit123 we need to sell the idea to SIG API Machinery and that could take time or even not happen at all.
A few unnecessary `String()` calls can be omitted, but I can live with them.
I give you LGTM on the kubeadm side of things.
/lgtm

**Comment by bart0sh** [comment]
@rosti, @neolit123 Thank you for the review!

As for the change of Version API, I tried to avoid it in the previous version of this patch. Then @fabriziopandini suggested this approach and I reworked the patch. I still have previous version, so please suggest which way we want to go.

In my opinion we should go faster way(previous patch) and submit Version API change as a separate PR. When/if it's accepted I'll be happy to submit new PR for kubeadm code that uses changed API.

Thoughts?

**Review by fabriziopandini — APPROVED** [review]
@bart0sh thanks for this news version!
/approve
/lgtm

Let's wait for sig-api-machinery opinion

**Comment by bart0sh** [comment]
@fabriziopandini @neolit123 Is anything required from me? Should I contact someone or should we assign someone from sig-api-machinery to this PR?

**Comment by neolit123** [comment]
@bart0sh the likelyhood of this making it in before NY is minimal due to the api-machinery change.

/assign @caesarxuchao

**Comment by bart0sh** [comment]
@neolit123 so, should we wait or use previous version of this PR and make separate PR for api change as I proposed earlier?

**Comment by neolit123** [comment]
@bart0sh 

as per @fabriziopandini 's comment:
>Let's wait for sig-api-machinery opinion

File: cmd/kubeadm/app/cmd/alpha/kubelet.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/alpha/kubelet.go]: @fabriziopandini @rosti @bart0sh @rojkov

instead of the usage of this constant here i think we should just have a `CurrentKubernetesMinorVersion` (or a similar naming) that is `v1.13.0`.
we can then use this version in such CLI examples and also have it as the current version for unit tests.
WDYT?

Comment (liggitt) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: would that `Current...` constant require manual updating on every release? that does not seem like a good idea

Comment (neolit123) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: we already bump these on each release:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L380-L383

in an ideal world:
- an app runtime should fail if the build system did not add versioning to the binary.
- skew versions and current versions should be obtained in variables based of the client version.
- unit tests should not be bound to internet or client versioning functionality.
- a default version should not be fetched from the internet by default.

Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: Considering that this PR IMO is already achieving the goals to simplify version bumps, which was the original scope of the underlying issue, I'm not sure it is a good idea to conflate in this PR other goals like fixing version discovery for unit test or shifting to a dynamic approach for handling version skew policy. Might be follow up PRs are more appropriate in this case

Comment (neolit123) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: then again, using `constants.MinimumKubeletVersion` as an example in the CLI might introduce problems. this is a version that is currently 1.12 instead of 1.13.

Comment (rosti) [reply cmd/kubeadm/app/cmd/alpha/kubelet.go]: The idea of `CurrentKubernetesVersion` is good one, but outside of the scope of this PR. It can be done in some of the followup PRs though. For a start we can bump this version manually at the end of the release cycle, but ideally this should be done automatically by the build system.

**Comment by rosti** [comment]
I think, that we can safely cancel the hold now and let this merge.
@neolit123 @fabriziopandini WDYT?

**Comment by bart0sh** [comment]
in my opinion it's ok to merge it. If there are any issues with it I'll fix it. It's reviewed and approved by many people already.

**Pull request merged**
